File: curriculum/18_search_engine/node-impl/tests/search.test.ts
import { describe, it, expect } from "vitest";
import { tokenize, InvertedIndex } from "../src/search.js";

describe("tokenize", () => {
  it("tokenizes basic text", () => {
    const tokens = tokenize("Hello, World! This is a test.");
    expect(tokens).toEqual(["hello", "world", "test"]);
  });

  it("removes stop words", () => {
    const tokens = tokenize("the quick brown fox jumps over the lazy dog");
    expect(tokens).not.toContain("the");
    expect(tokens).not.toContain("over");
    expect(tokens).toHaveLength(6);
  });

  it("handles empty string", () => {
    expect(tokenize("")).toEqual([]);
  });

  it("preserves numbers", () => {
    const tokens = tokenize("test123 word456");
    expect(tokens).toEqual(["test123", "word456"]);
  });
});

describe("InvertedIndex", () => {
  it("indexes and searches documents", () => {
    const idx = new InvertedIndex();
    idx.addDocument("Go Guide", "Go is a programming language designed for concurrency");
    idx.addDocument("Rust Guide", "Rust is a systems programming language focused on safety");
    idx.addDocument("Python Guide", "Python is a versatile programming language");

    const results = idx.search("programming language", 10);
    expect(results).toHaveLength(3);
  });

  it("ranks by TF-IDF (more occurrences = higher score)", () => {
    const idx = new InvertedIndex();
    idx.addDocument("Doc A", "rust rust rust programming");
    idx.addDocument("Doc B", "rust programming");

    const results = idx.search("rust", 10);
    expect(results).toHaveLength(2);
    expect(results[0].title).toBe("Doc A");
  });

  it("returns empty for non-matching query", () => {
    const idx = new InvertedIndex();
    idx.addDocument("Doc", "hello world");
    expect(idx.search("nonexistent", 10)).toEqual([]);
  });

  it("returns empty for empty query", () => {
    const idx = new InvertedIndex();
    idx.addDocument("Doc", "hello world");
    expect(idx.search("", 10)).toEqual([]);
  });

  it("respects limit", () => {
    const idx = new InvertedIndex();
    for (let i = 0; i < 10; i++) {
      idx.addDocument("Doc", "common word here");
    }
    const results = idx.search("common", 3);
    expect(results).toHaveLength(3);
  });

  it("tracks document count", () => {
    const idx = new InvertedIndex();
    expect(idx.documentCount).toBe(0);
    idx.addDocument("A", "content a");
    idx.addDocument("B", "content b");
    expect(idx.documentCount).toBe(2);
  });
});

File: curriculum/18_search_engine/rust-impl/tests/search_test.rs
use search_engine::{tokenize, InvertedIndex};

#[test]
fn test_tokenize_basic() {
    let tokens = tokenize("Hello, World! This is a test.");
    assert_eq!(tokens, vec!["hello", "world", "test"]);
}

#[test]
fn test_tokenize_stop_words() {
    let tokens = tokenize("the quick brown fox jumps over the lazy dog");
    assert!(!tokens.contains(&"the".to_string()));
    assert!(!tokens.contains(&"over".to_string()));
    assert_eq!(tokens.len(), 6);
}

#[test]
fn test_tokenize_empty() {
    let tokens = tokenize("");
    assert!(tokens.is_empty());
}

#[test]
fn test_tokenize_numbers_preserved() {
    let tokens = tokenize("test123 word456");
    assert_eq!(tokens.len(), 2);
    assert_eq!(tokens[0], "test123");
}

#[test]
fn test_index_and_search() {
    let mut idx = InvertedIndex::new();
    idx.add_document("Go Guide", "Go is a programming language designed for concurrency");
    idx.add_document("Rust Guide", "Rust is a systems programming language focused on safety");
    idx.add_document("Python Guide", "Python is a versatile programming language");

    let results = idx.search("programming language", 10);
    assert_eq!(results.len(), 3);
}

#[test]
fn test_search_ranking() {
    let mut idx = InvertedIndex::new();
    idx.add_document("Doc A", "rust rust rust programming");
    idx.add_document("Doc B", "rust programming");

    let results = idx.search("rust", 10);
    assert_eq!(results.len(), 2);
    assert_eq!(results[0].title, "Doc A"); // more occurrences → higher score
}

#[test]
fn test_search_not_found() {
    let mut idx = InvertedIndex::new();
    idx.add_document("Doc", "hello world");
    let results = idx.search("nonexistent", 10);
    assert!(results.is_empty());
}

#[test]
fn test_search_empty_query() {
    let mut idx = InvertedIndex::new();
    idx.add_document("Doc", "hello world");
    let results = idx.search("", 10);
    assert!(results.is_empty());
}

#[test]
fn test_search_limit() {
    let mut idx = InvertedIndex::new();
    for _ in 0..10 {
        idx.add_document("Doc", "common word here");
    }
    let results = idx.search("common", 3);
    assert_eq!(results.len(), 3);
}

#[test]
fn test_document_count() {
    let idx = InvertedIndex::new();
    assert_eq!(idx.document_count(), 0);
}
